pkg/artifactory/resource_artifactory_certificate.go

```diff
@@ -0,0 +1,194 @@
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"alias": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"content": {
+				Type:      schema.TypeString,
+				Required:  true,
+				Sensitive: true,
+			},
+			"fingerprint": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"issued_by": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"issued_on": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"issued_to": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"valid_until": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+
+		CustomizeDiff: func(d *schema.ResourceDiff, v interface{}) error {
+			fingerprint, err := calculateFingerPrint(d.Get("content").(string))
+			if err != nil {
+				return err
+			}
+			if d.Get("fingerprint").(string) != fingerprint {
+				d.SetNewComputed("fingerprint")
+			}
+			return nil
+		},
+	}
+}
+
+func formatFingerPrint(f []byte) string {
+	buf := make([]byte, 0, 3*len(f))
+	x := buf[1*len(f) : 3*len(f)]
+	hex.Encode(x, f)
+	for i := 0; i < len(x); i += 2 {
+		buf = append(buf, x[i], x[i+1], ':')
+	}
+	return strings.ToUpper(string(buf[:len(buf)-1]))
+}
+
+func extractCertificate(pemData string) (*x509.Certificate, error) {
+	block, rest := pem.Decode([]byte(pemData))
+	for block != nil {
+		if block.Type != "CERTIFICATE" {
+			block, rest = pem.Decode(rest)
+			continue
+		}
+
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return nil, err
+		}
+
+		return cert, nil
+	}
+
+	return nil, fmt.Errorf("no certificate in PEM data")
+}
+
+func calculateFingerPrint(pemData string) (string, error) {
+	cert, err := extractCertificate(pemData)
+	if err != nil {
+		return "", err
+	}
+
+	fingerprint := sha256.Sum256(cert.Raw)
+
+	return formatFingerPrint(fingerprint[:]), nil
+}
+
+func findCertificate(d *schema.ResourceData, m interface{}) (*v1.CertificateDetails, error) {
+	c := m.(*artifactory.Artifactory)
+
+	certs, _, err := c.V1.Security.GetCertificates(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	// No way other than to loop through each certificate
+	for _, cert := range *certs {
+		if *cert.CertificateAlias == d.Id() {
+			return &cert, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func resourceCertificateCreate(d *schema.ResourceData, m interface{}) error {
+	d.SetId(d.Get("alias").(string))
+	return resourceCertificateUpdate(d, m)
+}
+
+func resourceCertificateRead(d *schema.ResourceData, m interface{}) error {
+	cert, err := findCertificate(d, m)
+	if err != nil {
+		return err
+	}
+
+	if cert != nil {
+		d.Set("alias", *cert.CertificateAlias)
+		d.Set("fingerprint", *cert.FingerPrint)
+		d.Set("issued_by", *cert.IssuedBy)
+		d.Set("issued_on", *cert.IssuedOn)
+		d.Set("issued_to", *cert.IssuedTo)
+		d.Set("valid_until", *cert.ValidUntil)
```

Comment:
Could errors here be logged and returned too?

Reply:
Done.